### PR TITLE
Finalizing optimized HC set

### PR DESCRIPTION
**Pull request by tanyaveksler**
Initial implementation of building and minimizing fw-rules directly from connectivity properties.

### CODE REVIEW

File: nca/NetworkConfig/NetworkConfigQuery.py [hunk 1]
```diff
@@ -157,18 +158,28 @@ def compare_fw_rules(fw_rules1, fw_rules2, peer_container, rules_descr=""):
...
-            diff_prop = (conn_props1 - conn_props2) | (conn_props2 - conn_props1)
+            diff_prop = (props1 - props2) | (props2 - props1)
             if diff_prop.are_auto_conns():
                 print(f"{text_prefix} differ only in auto-connections")
             else:
                 print(f"Error: {text_prefix} are different")
                 assert False
 
+    @staticmethod
+    def compare_fw_rules_to_conn_props(fw_rules, props, peer_container, connectivity_restriction=None):
```
[the remark this answers is not in the document]
Comment (tanyaveksler) [reply nca/NetworkConfig/NetworkConfigQuery.py:178]: Checked, it runs both on `-opt=debug` and `-opt=true`.

File: nca/CoreDS/ConnectivityProperties.py [hunk 2]
```diff
@@ -496,9 +496,64 @@ def props_without_auto_conns(self):
...
+        """
         peers = self.project_on_one_dimension("src_peers") | self.project_on_one_dimension("dst_peers")
         auto_conns = ConnectivityProperties()
         for peer in peers:
             auto_conns |= ConnectivityProperties.make_conn_props_from_dict({"src_peers": PeerSet({peer}),
                                                                             "dst_peers": PeerSet({peer})})
-        return self - auto_conns
+        return auto_conns
+
+    def minimize(self):
```
[the remark this answers is not in the document]
Comment (tanyaveksler) [reply nca/CoreDS/ConnectivityProperties.py:513]: Checked. There are a few examples where it improves. For example,
`tests\istio_testcases\example_policies\bookinfo-demo\sidecar_examples\bookinfo-sidecar-tests-scheme.yaml`
and
`tests\istio_testcases\example_policies\bookinfo-demo\sidecar_examples\bookinfo-test-sidecar-connectivity-scheme.yaml`

File: nca/FWRules/MinimizeCsFWRulesOpt.py [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+        compute the union (set) of all peer pairs for which communication is allowed in current connection-set (but
+        not necessarily only limited to current connection set)
+        :return: None
+        """
+        covered_peer_props = self.peer_props | self.peer_props_in_containing_connections
+        all_peers_set = self.peer_props.get_all_peers()
+        for pod in all_peers_set:
+            if isinstance(pod, ClusterEP):
+                covered_peer_props |= ConnectivityProperties.make_conn_props_from_dict({"src_peers": PeerSet({pod}),
+                                                                                        "dst_peers": PeerSet({pod})})
```
Comment (adisos) [inline nca/FWRules/MinimizeCsFWRulesOpt.py:130]: this captures only pairs where both src and dst are `ClusterEP` ?
in the original version it was pairs where at lease one of these elements was  `ClusterEP`
```
        all_pods_set = set(src for (src, dst) in self.peer_pairs if isinstance(src, ClusterEP)) | \
            set(dst for (src, dst) in self.peer_pairs if isinstance(dst, ClusterEP))
        for pod in all_pods_set:
            covered_peer_pairs_union |= {(pod, pod)}
```

File: nca/FWRules/MinimizeCsFWRulesOpt.py [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+    def _add_all_fw_rules(self):
+        """
+        Computation of fw-rules, following the ns-grouping of peer_pairs.
+        Results are at: self.minimized_rules_set
+        :return: None
+        """
+        # create initial fw-rules from ns_set_pairs, base_elem_pairs, peer_props_without_ns_expr
+        initial_fw_rules = self._create_all_initial_fw_rules()
+        self.minimized_fw_rules = initial_fw_rules
+        return  # Tanya: temp
```
Comment (adisos) [inline nca/FWRules/MinimizeCsFWRulesOpt.py:470]: remove the code below

File: nca/FWRules/MinimizeCsFWRulesOpt.py (hunk 3 above)
Comment (tanyaveksler) [reply nca/FWRules/MinimizeCsFWRulesOpt.py:130]: It calculates all `ClusterEP peers` that appear in connections. The original calculation took them from `peer_pairs` (taking `src` if `src` is `ClusterEP` or `dst` if `dst` is `ClusterEP`).
The optimized calculation first computes `all_peers_set` by taking the union of projection of connections to `src_peers` and to `dst_peers`, and then takes from them those peers that are `ClusterEP`.
Thus, the both original and optimized calculations are equivalent.

File: nca/FWRules/MinimizeCsFWRulesOpt.py (hunk 4 above)
Comment (tanyaveksler) [reply nca/FWRules/MinimizeCsFWRulesOpt.py]: Done

File: nca/FWRules/MinimizeCsFWRulesOpt.py (hunk 3 above)
Comment (adisos) [reply nca/FWRules/MinimizeCsFWRulesOpt.py:132]: but in this version both src and dst are `ClusterEP` , and in the original version only one of src or dst?

**Pull request merged**